### PR TITLE
Parsing Inline References

### PR DESCRIPTION
**Pull request by stathischaritos**
Connects to #44
We parse the inline references and add them in a list inside ParagraphsWithLinks. We keep the inner text ,start and end  position, to be used by wikistats for ignoring them.

### CODE REVIEW

**Comment by stathischaritos** [comment]
Issue #44

File: src/main/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParser.java [hunk 1]
```diff
@@ -633,4 +637,24 @@ private boolean isImage(de.tudarmstadt.ukp.wikipedia.parser.Link link)
...
+     * Returns a list with all the inline references found in a piece of text.
+     * @param text - A piece of text that contains inline references.
+     */
+    private List<Ref> extractInlineReferences(String text)
+    {
+        Matcher m = refsPattern.matcher(text.toString());
+        List<Ref> refs = new ArrayList();
+
+        while(m.find()) {
+            // Replace with whitespace.
```
Comment (tgalery) [inline src/main/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParser.java]: I guess this comment is a bit misleading ?

File: src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java [hunk 2]
```diff
@@ -123,6 +123,19 @@ public void testExtractingLinksOtherLang(){
 
 	}
 
+    /*
+    * Test for inline ref parsing.
+    * */
+	@Test
+	public void testReferences() throws IOException {
```
Comment (tgalery) [inline src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java]: Might be good to add more tests, e.g. expectations about that's the text of the reference and stuff like this.  Also it seems that the this article has lots of references, so we might have expectations about the number of references extracted and so on ?

**Comment by tgalery** [comment]
I think it would be good to squash all the commits you have into one after you address comments and maybe we can run this on the "simple" wikipedia

File: src/test/resources/misc/anarchism.xml [hunk 3]
```diff
@@ -0,0 +1,453 @@
+<mediawiki xmlns="http://www.mediawiki.org/xml/export-0.10/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mediawiki.org/xml/export-0.10/ http://www.mediawiki.org/xml/export-0.10.xsd" version="0.10" xml:lang="en">
```
Comment (tgalery) [inline src/test/resources/misc/anarchism.xml]: Also, wouldn't it be more consistent to add just the page markup (without the top level xml stuff) into `resources/en` instead of `resources/misc` ?

File: src/main/java/it/cnr/isti/hpc/wikipedia/article/Ref.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package it.cnr.isti.hpc.wikipedia.article;
+
+/**
+ *  This class represents inline wikipedia references. You can find the pattern to detect those inside ArticleParser.
```
Comment (hmcc) [inline src/main/java/it/cnr/isti/hpc/wikipedia/article/Ref.java]: There's a way to link to other classes in Javadoc, something like `{@link ArticleParser}`, which renders the link clickable when the Javadoc is processed.

File: src/main/java/it/cnr/isti/hpc/wikipedia/article/Ref.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public String getText() {
+        return text;
+    }
+
+    public void setText(String text) {
+        this.text = text;
+    }
+
```
Comment (hmcc) [inline src/main/java/it/cnr/isti/hpc/wikipedia/article/Ref.java]: 👾

File: src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java [hunk 6]
```diff
@@ -1,50 +1,50 @@
...
+import java.io.IOException;
 import java.io.InputStreamReader;
 import java.util.ArrayList;
 import java.util.List;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertThat;
 import static org.hamcrest.CoreMatchers.hasItems;
+import static org.junit.Assert.*;
```
Comment (hmcc) [inline src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java]: avoid star imports please

File: src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java [hunk 7]
```diff
@@ -124,13 +124,39 @@ public void testExtractingLinksOtherLang(){
...
+		Article a = new Article();
+		String mediawiki = IOUtils.getFileAsUTF8String("./src/test/resources/misc/anarchism.xml");
+		parser.parse(a, mediawiki);
+
+		// check that the first paragraph of anarchism has 10 refs detected.
+		List<Ref> refs = a.getParagraphsWithLinks().get(0).getRefs();
+		assertEquals(10, refs.size());
+
+		// Check first and last detected references inside values
+		Ref fistRef = refs.get(0);
```
Comment (hmcc) [inline src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java]: firstRef?

File: src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java [hunk 8]
```diff
@@ -124,13 +124,39 @@ public void testExtractingLinksOtherLang(){
...
+		assertEquals(157, fistRef.getStart());
+		assertEquals(435, fistRef.getEnd());
+		assertNotEquals(-1, fistRef.getText().indexOf("a social philosophy that rejects authoritarian government and maintains"));
```
Comment (hmcc) [inline src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java]: This reads a bit funny, like a double negative or something.
What about
```
assertTrue(firstRef.contains("a social philosophy that rejects authoritarian government and maintains"))
```
?

Comment (hmcc) [reply src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java]: You could also refactor these 4 lines into a helper method like
```
checkRef(ref, start, end, text) {
    assertEquals(start, ref.getStart());
    assertEquals(end, ref.getEnd());
    assertTrue(ref.contains(text))
}
```
which would allow people to extend the tests more easily in future.

File: src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java [hunk 9]
```diff
@@ -124,13 +124,39 @@ public void testExtractingLinksOtherLang(){
...
+
+		Ref lastRef = refs.get(9);
+		assertEquals(2529, lastRef.getStart());
+		assertEquals(2713, lastRef.getEnd());
+		assertNotEquals(-1, lastRef.getText().indexOf("The Concise Oxford Dictionary of Politics. Ed. Iain McLean and Alistair McMillan"));
+	}
```
Comment (hmcc) [inline src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java]: 👍 for the test, but it would be more unit-test-like to make the `extractInlineReferences` method protected and then test it directly.

Some more fun things to test:
* `extractInlineReferences(null)`
* `extractInlineReferences("")`
* `extractInlineReferences("  ")`

**Comment by hmcc** [comment]
Minor nitpicks, but looking good!

File: src/main/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParser.java [hunk 10]
```diff
@@ -635,16 +635,16 @@ private boolean isImage(de.tudarmstadt.ukp.wikipedia.parser.Link link)
...
     {
-        Matcher m = refsPattern.matcher(text.toString());
-        List<Ref> refs = new ArrayList();
-
-        while(m.find()) {
+    	List<Ref> refs = new ArrayList();
+		if (text == null || text.isEmpty()) return refs;
+		Matcher m = refsPattern.matcher(text.toString());
+		while(m.find()) {
             int start = m.start();
```
Comment (hmcc) [inline src/main/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParser.java]: indentation 😞

File: src/test/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParserTest.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    /*
+	* Nested reference extraction.
+	* */
+    @Test
+    public void testNestedReferences() throws IOException {
+        ArticleParser parser = new ArticleParser(Language.EN);
+        assertEquals(1, parser.extractInlineReferences("Before <ref>Hello World <ref name=&quote;nested&quote;>Hello Again World</ref> Still not out of the top reference! </ref> After.").size());
+    }
+    /*
+	* More reference parsing edge cases.
```
Comment (hmcc) [inline src/test/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParserTest.java]: indentation

File: src/test/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParserTest.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+	* Basic reference extraction check.
+	* */
+    @Test
+    public void testBasicReferences() throws IOException {
+        ArticleParser parser = new ArticleParser(Language.EN);
+        List<Ref> refs = parser.extractInlineReferences("Before <ref>Hello World</ref> After.");
+        assertEquals(1, refs.size());
+    }
+    /*
+	* Nested reference extraction.
```
Comment (hmcc) [inline src/test/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParserTest.java]: indentation

File: src/test/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParserTest.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class ArticleParserTest {
+    /*
+	* Basic reference extraction check.
```
Comment (hmcc) [inline src/test/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParserTest.java]: indentation

File: src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java [hunk 14]
```diff
@@ -138,15 +149,12 @@ public void testReferences() throws IOException {
...
-		assertNotEquals(-1, fistRef.getText().indexOf("a social philosophy that rejects authoritarian government and maintains"));
+		Ref firstRef = refs.get(0);
+		checkRef(firstRef, 157, 435, "a social philosophy that rejects authoritarian government and maintains");
 
 		Ref lastRef = refs.get(9);
-		assertEquals(2529, lastRef.getStart());
-		assertEquals(2713, lastRef.getEnd());
-		assertNotEquals(-1, lastRef.getText().indexOf("The Concise Oxford Dictionary of Politics. Ed. Iain McLean and Alistair McMillan"));
+		checkRef(lastRef, 2529, 2713, "The Concise Oxford Dictionary of Politics. Ed. Iain McLean and Alistair McMillan");
+
```
Comment (hmcc) [inline src/test/java/it/cnr/isti/hpc/wikipedia/article/ArticleTest.java]: 👾
